### PR TITLE
Expose d-bus method to trigger panel functions

### PR DESCRIPTION
**Pull request by PriyangaRamasamy**
This commit provides support for a d-bus method to trigger IBMi functions on panel via d-bus call.

The method "TriggerPanelFunction" can be found under "com.ibm.panel" interface which accepts only the function numbers which are IBMi functions.

Supported functions are{21, 22, 34, 65, 66, 67,68, 69, 70}.

Test:
:~# busctl call com.ibm.PanelApp /com/ibm/panel_app com.ibm.panel ExecuteFunction y 30
(bss) false "" ""
  Jul 27 06:11:20 p10bmc pldmd[1768]: 80 02 0a 01 01 00 1b 00 01 00 02 00

:~# busctl call com.ibm.PanelApp /com/ibm/panel_app com.ibm.panel ExecuteFunction y 21
(bss) true "21    00" ""
  Jul 27 06:25:06 p10bmc ibm-panel[1798]: packet data sent to pldm: 80 02 39 08 00 01 01 15

:~# busctl call com.ibm.PanelApp /com/ibm/panel_app com.ibm.panel ExecuteFunction y 22
(bss) true "22    00" ""
 Jul 27 06:26:30 p10bmc ibm-panel[1798]: packet data sent to pldm: 80 02 39 08 00 01 01 16

Change-Id: I2450cfda8341aa7c1546d8fa477a8c8fa35c56f8

### CODE REVIEW

**Comment by PriyangaRamasamy** [comment]
jenkins run tests please

File: include/bus_handler.hpp [hunk 1]
```diff
@@ -65,6 +65,10 @@ class BusHandler
                                      }
                                      return 0;
                                  });
+        iface->register_method("TriggerPanelFunction",
```
Comment (SunnySrivastava1984) [inline include/bus_handler.hpp:68]: These functions have nothing to do with panel. 
Please expose a generic name - "TriggerFunction".

**Review by SunnySrivastava1984 — CHANGES_REQUESTED** [review]

File: src/bus_handler.cpp [hunk 2]
```diff
@@ -81,4 +81,23 @@ void BusHandler::btnRequest(int event)
...
+    const std::vector<types::FunctionNumber> supportedFunctions = {
+        21, 22, 34, 65, 66, 67, 68, 69, 70};
+
+    if (std::find(supportedFunctions.begin(), supportedFunctions.end(),
+                  funcNum) == std::end(supportedFunctions))
+    {
+        std::cout << "The given function number " << static_cast<int>(funcNum)
+                  << " is supported only to trigger via lcd panel buttons. No "
+                     "support to trigger via d-bus."
+                  << std::endl;
```
[the remark this answers is not in the document]
Comment (PriyangaRamasamy) [reply src/bus_handler.cpp:96]: this is just a trace to tell that the input is unsupported. if the user triggers it from command line, the trace can be identified from journal only. 
from GUI point of view, this is going to be similar to how lamptest request works. it's going to be a button press (for the above functions) without expecting any ack from panel app.

File: include/bus_handler.hpp (hunk 1 above)
Comment (jinuthomas) [reply include/bus_handler.hpp:68]: better would be executeFunction

File: src/bus_handler.cpp [hunk 3]
```diff
@@ -81,4 +81,23 @@ void BusHandler::btnRequest(int event)
...
+        return;
```
Comment (GiridhariKrishna) [inline src/bus_handler.cpp:97]: cerr should be good to address, here.

File: include/bus_handler.hpp [hunk 4]
```diff
@@ -116,6 +120,13 @@ class BusHandler
      * @param[in] event: Button event
      */
     void btnRequest(int event);
+
+    /**
+     * @brief API to trigger specific panel functions.
+     *
+     * @param[in] funcNum - Function number to execute
+     */
+    void triggerPanelFunc(const types::FunctionNumber funcNum);
```
Comment (GiridhariKrishna) [inline include/bus_handler.hpp:129]: It can be used for all functions that are valid to be executed, so may be a check or valid supported functions in description will be good.

File: src/bus_handler.cpp (hunk 3 above)
Comment (PriyangaRamasamy) [reply src/bus_handler.cpp]: done

File: src/bus_handler.cpp [hunk 5]
```diff
@@ -81,4 +81,23 @@ void BusHandler::btnRequest(int event)
...
+
+void BusHandler::triggerPanelFunc(const types::FunctionNumber funcNum)
+{
+    const std::vector<types::FunctionNumber> supportedFunctions = {
+        21, 22, 34, 65, 66, 67, 68, 69, 70};
+
+    if (std::find(supportedFunctions.begin(), supportedFunctions.end(),
+                  funcNum) == std::end(supportedFunctions))
+    {
+        std::cout << "The given function number " << static_cast<int>(funcNum)
```
[the remark this answers is not in the document]
Comment (PriyangaRamasamy) [reply src/bus_handler.cpp]: done

File: src/bus_handler.cpp [hunk 6]
```diff
@@ -81,4 +81,23 @@ void BusHandler::btnRequest(int event)
             std::cerr << "Invalid Input" << std::endl;
     }
 }
+
+void BusHandler::triggerPanelFunc(const types::FunctionNumber funcNum)
```
[the remark this answers is not in the document]
Comment (PriyangaRamasamy) [reply src/bus_handler.cpp]: done

File: include/bus_handler.hpp (hunk 4 above)
Comment (PriyangaRamasamy) [reply include/bus_handler.hpp]: done

File: include/bus_handler.hpp [hunk 7]
```diff
@@ -116,6 +120,13 @@ class BusHandler
      * @param[in] event: Button event
      */
     void btnRequest(int event);
+
+    /**
+     * @brief API to trigger specific panel functions.
```
[the remark this answers is not in the document]
Comment (PriyangaRamasamy) [reply include/bus_handler.hpp]: done

File: include/bus_handler.hpp (hunk 1 above)
Comment (PriyangaRamasamy) [reply include/bus_handler.hpp]: done

File: src/bus_handler.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (PriyangaRamasamy) [reply src/bus_handler.cpp]: returning boolean to the caller

**Comment by PriyangaRamasamy** [comment]
jenkins run tests please

File: src/bus_handler.cpp [hunk 8]
```diff
@@ -81,4 +81,22 @@ void BusHandler::btnRequest(int event)
...
+
+    if (std::find(supportedFunctions.begin(), supportedFunctions.end(),
+                  funcNum) == std::end(supportedFunctions))
+    {
+        std::cerr << "Function number " << static_cast<int>(funcNum)
+                  << " is unsupported to execute via d-bus." << std::endl;
+        return false;
+    }
+
+    executor->executeFunction(funcNum, types::FunctionalityList{});
```
Comment (jinuthomas) [inline src/bus_handler.cpp:98]: I am not sure this will work

File: include/bus_handler.hpp [hunk 9]
```diff
@@ -65,6 +65,10 @@ class BusHandler
                                      }
                                      return 0;
                                  });
+        iface->register_method("ExecuteFunction",
```
Comment (jinuthomas) [inline include/bus_handler.hpp:68]: do we need to catch exceptions ?

**Review by jinuthomas — CHANGES_REQUESTED** [review]

File: include/bus_handler.hpp (hunk 9 above)
Comment (PriyangaRamasamy) [reply include/bus_handler.hpp:68]: currently am catching the exception and returning false.

File: src/bus_handler.cpp [hunk 10]
```diff
@@ -81,4 +81,22 @@ void BusHandler::btnRequest(int event)
...
+bool BusHandler::triggerPanelFunc(const types::FunctionNumber funcNum)
+{
+    const std::vector<types::FunctionNumber> supportedFunctions = {
+        21, 22, 34, 65, 66, 67, 68, 69, 70};
+
+    if (std::find(supportedFunctions.begin(), supportedFunctions.end(),
+                  funcNum) == std::end(supportedFunctions))
+    {
+        std::cerr << "Function number " << static_cast<int>(funcNum)
+                  << " is unsupported to execute via d-bus." << std::endl;
```
[the remark this answers is not in the document]
Comment (PriyangaRamasamy) [reply src/bus_handler.cpp]: removed

File: src/executor.cpp [hunk 11]
```diff
@@ -1038,4 +1040,52 @@ void Executor::execute74()
...
+}
+
+void Executor::storeDbusExecStatus(bool returnState, std::string line1,
+                                   std::string line2)
+{
+    if (!isThisReqFromDbus)
+    {
+        return;
+    }
+    std::get<0>(executionStatus) = returnState;
```
Comment (SunnySrivastava1984) [inline src/executor.cpp:1087]: std::make_tuple(returnState, line1, line2);

just one line, will make the code look cleaner.

File: src/executor.cpp [hunk 12]
```diff
@@ -30,6 +30,8 @@ void Executor::displayExecutionStatus(
     }
 
     convert << (result ? " 00" : " FF");
+
+    storeDbusExecStatus(result, convert.str(), "");
```
Comment (SunnySrivastava1984) [inline src/executor.cpp:34]: Why this function is required? You have a class member variable for this just call std::make_tuple with the values and assign that in place.

File: src/executor.cpp [hunk 13]
```diff
@@ -1038,4 +1040,52 @@ void Executor::execute74()
...
+        storeDbusExecStatus(false, "", "");
+        std::cerr << e.what() << std::endl;
+        return executionStatus;
+    }
+
+    isThisReqFromDbus = false;
+    return executionStatus;
+}
+
+void Executor::storeDbusExecStatus(bool returnState, std::string line1,
```
Comment (SunnySrivastava1984) [inline src/executor.cpp:1080]: I dont see need of this method.

File: src/executor.cpp [hunk 14]
```diff
@@ -1038,4 +1040,52 @@ void Executor::execute74()
...
+
+            default:
+                std::runtime_error("Given function number is not supported to "
+                                   "execute via d-bus.");
+        }
+    }
+    catch (std::exception& e)
+    {
+        isThisReqFromDbus = false;
+        storeDbusExecStatus(false, "", "");
```
Comment (SunnySrivastava1984) [inline src/executor.cpp:1071]: Call the class level variable and assign values in-place

**Review by SunnySrivastava1984 — CHANGES_REQUESTED** [review]

File: src/executor.cpp (hunk 14 above)
Comment (PriyangaRamasamy) [reply src/executor.cpp]: done

File: src/executor.cpp (hunk 13 above)
Comment (PriyangaRamasamy) [reply src/executor.cpp]: removed

File: src/executor.cpp [hunk 15]
```diff
@@ -1038,4 +1040,52 @@ void Executor::execute74()
...
+            case 66:
+            case 67:
+            case 68:
+            case 69:
+            case 70:
+                sendFuncNumToPhyp(funcNum);
+                break;
+
+            default:
+                std::runtime_error("Given function number is not supported to "
```
[the remark this answers is not in the document]
Comment (PriyangaRamasamy) [reply src/executor.cpp]: not really required. have changed it

File: include/bus_handler.hpp (hunk 9 above)
Comment (SunnySrivastava1984) [inline include/bus_handler.hpp:68]: nit: Please have a blank line between two blocks of code.

File: include/bus_handler.hpp [hunk 16]
```diff
@@ -116,6 +120,18 @@ class BusHandler
...
     void btnRequest(int event);
+
+    /**
+     * @brief API to execute panel function number.
+     *
+     * Method to trigger panel function via d-bus. Currently only certain
+     * functions are supported to trigger using d-bus calls.
+     *
+     * @param[in] funcNum - Function number to execute
+     * @return true if execution is triggered, false otherwise.
```
Comment (SunnySrivastava1984) [inline include/bus_handler.hpp:131]: Returns a tuple. Please mention that and also the value of fields in pass/fail condition. Like you have done in executor.hpp

File: include/bus_handler.hpp [hunk 17]
```diff
@@ -116,6 +120,18 @@ class BusHandler
...
+     */
+    types::ExecuteFunctionReturnType
```
Comment (SunnySrivastava1984) [inline include/bus_handler.hpp:133]: Rename it as "ExecutonResult"?

File: include/types.hpp [hunk 18]
```diff
@@ -69,6 +69,8 @@ using InterfacePropertyPair = std::pair<std::string, PropertyValueMap>;
 using singleObjectEntry = std::pair<sdbusplus::message::object_path,
                                     std::vector<InterfacePropertyPair>>;
 
+using ExecuteFunctionReturnType = std::tuple<bool, std::string, std::string>;
```
Comment (SunnySrivastava1984) [inline include/types.hpp:72]: rename as "ExecutionResult"?

**Review by SunnySrivastava1984 — CHANGES_REQUESTED** [review]

**Review by SunnySrivastava1984 — APPROVED** [review]
Looks ok to me.

File: src/executor.cpp [hunk 19]
```diff
@@ -1038,4 +1045,41 @@ void Executor::execute74()
...
+                executionStatus = std::make_tuple(false, "", "");
+        }
+    }
+    catch (std::exception& e)
+    {
+        executionStatus = std::make_tuple(false, "", "");
+        std::cerr << e.what() << std::endl;
+    }
+
+    isExternallyTriggered = false;
```
Comment (GiridhariKrishnan) [inline src/executor.cpp:1082]: I don't see it used anywhere between true and false, so setting/inset is required?

**Review by GiridhariKrishnan — CHANGES_REQUESTED** [review]

File: src/panel_state_manager.cpp [hunk 20]
```diff
@@ -1095,6 +1095,29 @@ void PanelStateManager::setCEState()
...
+{
+    auto pos = find_if(panelFunctions.begin(), panelFunctions.end(),
+                       [funcNum](const PanelFunctionality& afunctionality) {
+                           return afunctionality.functionNumber == funcNum;
+                       });
+    if (pos != panelFunctions.end())
+    {
+        return pos->functionActiveState;
+    }
+    return false;
```
Comment (SunnySrivastava1984) [inline src/panel_state_manager.cpp:1109]: Just returning flag states only that function is not active, In case function number is not found, some error logging should be there.

File: src/utils.cpp [hunk 21]
```diff
@@ -80,8 +80,8 @@ void sendCurrDisplayToPanel(const std::string& line1, const std::string& line2,
     // Make these couts and other traces in the code configurable. Keeping
     // these commented until then as they flood the systemd journal.
     // couts are for debugging purpose. can be removed once the testing is done.
-    // std::cout << "L1 : " << line1 << std::endl;
-    // std::cout << "L2 : " << line2 << std::endl;
+     std::cout << "L1 : " << line1 << std::endl;
```
Comment (SunnySrivastava1984) [inline src/utils.cpp:83]: This is required?

File: src/executor.cpp (hunk 19 above)
Comment (SunnySrivastava1984) [reply src/executor.cpp:1082]: It is being used in function where we display.

**Review by SunnySrivastava1984 — CHANGES_REQUESTED** [review]

File: src/panel_state_manager.cpp (hunk 20 above)
Comment (PriyangaRamasamy) [reply src/panel_state_manager.cpp:1109]: we discussed this already, if the user given function is not valid, we send the execution status as false.. I think that is enough.

File: src/utils.cpp (hunk 21 above)
Comment (PriyangaRamasamy) [reply src/utils.cpp:83]: will remove it

File: src/panel_state_manager.cpp (hunk 20 above)
Comment (SunnySrivastava1984) [reply src/panel_state_manager.cpp:1109]: returning from DBus is Ok, my point is a log here to know when we return false for Dbus call, was it for some other error or function was disabled.

File: include/bus_handler.hpp [hunk 22]
```diff
@@ -116,6 +121,18 @@ class BusHandler
      * @param[in] event: Button event
      */
     void btnRequest(int event);
+
+    /**
+     * @brief API to execute panel function number.
+     *
+     * Method to trigger panel function via d-bus. Currently only certain
+     * functions are supported to trigger using d-bus calls.
```
Comment (jinuthomas) [inline include/bus_handler.hpp:129]: Remove Currently only certain
      * functions are supported to trigger using d-bus calls.

File: include/executor.hpp [hunk 23]
```diff
@@ -149,6 +149,18 @@ class Executor
...
+    /**
+     * @brief API to execute functions on requests from d-bus
+     * This method is called whenever there is an ExecuteFunction request from
+     * d-bus.
+     *
+     * @param[in] funcNum - Function number.
+     *
+     * @return Execution status(success/failure, display line 1, display line2)
+     */
+    types::FunctionExecutionResult
```
Comment (jinuthomas) [inline include/executor.hpp:161]: Would suggest RetunStatus  instead of FunctionExecutionResult

File: include/executor.hpp [hunk 24]
```diff
@@ -149,6 +149,18 @@ class Executor
...
     }
 
+    /**
+     * @brief API to execute functions on requests from d-bus
+     * This method is called whenever there is an ExecuteFunction request from
+     * d-bus.
+     *
+     * @param[in] funcNum - Function number.
+     *
+     * @return Execution status(success/failure, display line 1, display line2)
```
Comment (jinuthomas) [inline include/executor.hpp:159]: remove Execution

File: include/executor.hpp [hunk 25]
```diff
@@ -149,6 +149,18 @@ class Executor
...
+     */
+    types::FunctionExecutionResult
+        executeFuncViaDbus(const types::FunctionNumber funcNum);
```
Comment (jinuthomas) [inline include/executor.hpp:162]: executeFunctionDirectly

File: include/executor.hpp [hunk 26]
```diff
@@ -324,5 +336,10 @@ class Executor
     /* SRC and HEX words */
     std::string latestSrcAndHexwords;
 
+    /* To keep track if the execute request is from D-bus or not */
```
Comment (jinuthomas) [inline include/executor.hpp:339]: nit: external or not

File: include/executor.hpp [hunk 27]
```diff
@@ -324,5 +336,10 @@ class Executor
...
+    bool isExternallyTriggered = false;
+
+    types::FunctionExecutionResult executionStatus =
```
Comment (jinuthomas) [inline include/executor.hpp:342]: why this is required in the class ?

File: include/panel_state_manager.hpp [hunk 28]
```diff
@@ -136,6 +136,14 @@ class PanelStateManager
      */
     void setCEState();
 
+    /**
+     * @brief API to trigger functions on request from D-bus
+     * @param[in] funcNum - Function number
+     * @return Execution status of the function
+     */
+    types::FunctionExecutionResult
+        funcTriggerFromDbus(const types::FunctionNumber funcNum);
```
Comment (jinuthomas) [inline include/panel_state_manager.hpp:145]: triggerFunctionDirectly

File: include/panel_state_manager.hpp [hunk 29]
```diff
@@ -136,6 +136,14 @@ class PanelStateManager
      */
     void setCEState();
 
+    /**
+     * @brief API to trigger functions on request from D-bus
```
Comment (jinuthomas) [inline include/panel_state_manager.hpp:140]: trigger function directly

File: include/bus_handler.hpp [hunk 30]
```diff
@@ -116,6 +121,18 @@ class BusHandler
...
+     *
+     * @param[in] funcNum - Function number to execute
+     * @return a tuple with execution status(true/false) and display lines.
```
Comment (SunnySrivastava1984) [inline include/bus_handler.hpp:132]: Jinu: Remove execution from this line.

File: include/bus_handler.hpp [hunk 31]
```diff
@@ -116,6 +121,18 @@ class BusHandler
...
+     */
+    types::FunctionExecutionResult
```
Comment (SunnySrivastava1984) [inline include/bus_handler.hpp:134]: Jinu: Please refer my comment below.

File: include/executor.hpp [hunk 32]
```diff
@@ -149,6 +149,18 @@ class Executor
         latestSrcAndHexwords = pelEventId;
     }
 
+    /**
+     * @brief API to execute functions on requests from d-bus
```
Comment (SunnySrivastava1984) [inline include/executor.hpp:153]: Jinu: Remove DBus put external source.

File: include/executor.hpp [hunk 33]
```diff
@@ -149,6 +149,18 @@ class Executor
...
+     * This method is called whenever there is an ExecuteFunction request from
```
Comment (SunnySrivastava1984) [inline include/executor.hpp:154]: JInu: rephrase: This method is called whenever there is an external request to trigger a function.

File: include/panel_state_manager.hpp [hunk 34]
```diff
@@ -199,6 +207,13 @@ class PanelStateManager
      */
     void updateFunctionStatus();
 
+    /**
+     * @brief API which tells if the function is enabled or not
+     * @param[in] funcNum - Function number
+     * @return (1/0/-1) if function is enabled/disabled/invalid respectively.
```
Comment (SunnySrivastava1984) [inline include/panel_state_manager.hpp:213]: Previous implementation was ok. This function needs to just return if the given function number is enabled or not.

File: include/types.hpp [hunk 35]
```diff
@@ -69,6 +69,8 @@ using InterfacePropertyPair = std::pair<std::string, PropertyValueMap>;
 using singleObjectEntry = std::pair<sdbusplus::message::object_path,
                                     std::vector<InterfacePropertyPair>>;
 
+using FunctionExecutionResult = std::tuple<bool, std::string, std::string>;
```
Comment (SunnySrivastava1984) [inline include/types.hpp:72]: Jinu: Same as above, update the variable name.

File: src/executor.cpp [hunk 36]
```diff
@@ -30,6 +30,13 @@ void Executor::displayExecutionStatus(
     }
 
     convert << (result ? " 00" : " FF");
+
+    if (isExternallyTriggered)
+    {
+        executionStatus = std::make_tuple(result, convert.str(), "");
```
Comment (SunnySrivastava1984) [inline src/executor.cpp:36]: Return directly. No Need of variable?

File: src/executor.cpp [hunk 37]
```diff
@@ -1038,4 +1045,41 @@ void Executor::execute74()
...
+            case 68:
+            case 69:
+            case 70:
+                sendFuncNumToPhyp(funcNum);
+                break;
+
+            default:
+                std::cerr << "Given function number "
+                          << static_cast<int>(funcNum)
+                          << " is not supported to execute via d-bus."
```
Comment (SunnySrivastava1984) [inline src/executor.cpp:1071]: Jinu: Keep only till supported.

**Review by SunnySrivastava1984 — CHANGES_REQUESTED** [review]

File: src/executor.cpp (hunk 37 above)
Comment (PriyangaRamasamy) [reply src/executor.cpp:1071]: done

File: src/executor.cpp (hunk 36 above)
Comment (PriyangaRamasamy) [reply src/executor.cpp:36]: am not returning anything here.

File: include/panel_state_manager.hpp (hunk 34 above)
Comment (PriyangaRamasamy) [reply include/panel_state_manager.hpp:213]: done

File: include/executor.hpp (hunk 33 above)
Comment (PriyangaRamasamy) [reply include/executor.hpp:154]: done

File: include/executor.hpp (hunk 32 above)
Comment (PriyangaRamasamy) [reply include/executor.hpp:153]: done

File: include/bus_handler.hpp (hunk 31 above)
Comment (PriyangaRamasamy) [reply include/bus_handler.hpp:134]: done

File: include/bus_handler.hpp (hunk 30 above)
Comment (PriyangaRamasamy) [reply include/bus_handler.hpp:132]: done

File: src/panel_state_manager.cpp [hunk 38]
```diff
@@ -1095,6 +1095,43 @@ void PanelStateManager::setCEState()
...
+    // return -1 if the given function is invalid
+    return -1;
+}
+
+types::FunctionExecutionResult
+    PanelStateManager::funcTriggerFromDbus(const types::FunctionNumber funcNum)
+{
+    const auto funcStatus = isFunctionEnabled(funcNum);
+
+    if (funcStatus == 1)
```
[the remark this answers is not in the document]
Comment (PriyangaRamasamy) [reply src/panel_state_manager.cpp:1120]: removed

File: src/panel_state_manager.cpp [hunk 39]
```diff
@@ -1095,6 +1095,43 @@ void PanelStateManager::setCEState()
         }
     }
 }
+
+int PanelStateManager::isFunctionEnabled(const types::FunctionNumber funcNum)
```
[the remark this answers is not in the document]
Comment (PriyangaRamasamy) [reply src/panel_state_manager.cpp:1099]: done

File: include/panel_state_manager.hpp (hunk 29 above)
Comment (PriyangaRamasamy) [reply include/panel_state_manager.hpp:140]: done

File: include/panel_state_manager.hpp (hunk 28 above)
Comment (PriyangaRamasamy) [reply include/panel_state_manager.hpp:145]: done

File: include/bus_handler.hpp (hunk 22 above)
Comment (PriyangaRamasamy) [reply include/bus_handler.hpp]: done

File: include/executor.hpp (hunk 23 above)
Comment (PriyangaRamasamy) [reply include/executor.hpp]: done

File: include/executor.hpp (hunk 27 above)
Comment (PriyangaRamasamy) [reply include/executor.hpp]: This is to keep track of executionStatus without changing the signature of existing apis - executeFunction() and displayExecutionStatus()

File: src/executor.cpp (hunk 19 above)
Comment (GiridhariKrishna) [reply src/executor.cpp:1081]: Okay, then may be it should be unset there itself.

File: include/bus_handler.hpp [hunk 40]
```diff
@@ -65,6 +65,15 @@ class BusHandler
                                      }
                                      return 0;
                                  });
+
+        iface->register_method("ExecuteFunction",
+                               [this](const types::FunctionNumber funcNum) {
+                                   return this->triggerPanelFunc(funcNum);
+                               });
+
+        iface->register_method("getEnabledFunctions", [this]() {
```
Comment (jinuthomas) [inline include/bus_handler.hpp:74]: why this change? and also its corresponding changes ?

**Review by jinuthomas — CHANGES_REQUESTED** [review]

File: include/bus_handler.hpp (hunk 40 above)
Comment (PriyangaRamasamy) [reply include/bus_handler.hpp:74]: this is done as part of second commit in the same PR

Comment (jinuthomas) [reply include/bus_handler.hpp:74]: The Story was for completing the first part in this sprint , why the mixup ?

File: src/executor.cpp (hunk 19 above)
Comment (SunnySrivastava1984) [reply src/executor.cpp:1081]: What difference it will make? Set and unset in the same function will keep it easy to track.

File: src/panel_state_manager.cpp [hunk 41]
```diff
@@ -1095,6 +1095,52 @@ void PanelStateManager::setCEState()
...
+    const types::FunctionNumber funcNum)
+{
+    const auto funcStatus = isFunctionEnabled(funcNum);
+
+    if (funcStatus)
+    {
+        return (funcExecutor->executeFunctionDirectly(funcNum));
+    }
+
+    std::cerr << "Function " << static_cast<int>(funcNum) << " is disabled."
```
Comment (SunnySrivastava1984) [inline src/panel_state_manager.cpp:1123]: funcStatus variable not required.
Also, the log should move in the function "isFunctionEnabled".
So that all caller need not have this log (in case in future there are more callers to api "isFunctionEnabled")

File: src/panel_state_manager.cpp [hunk 42]
```diff
@@ -1095,6 +1095,52 @@ void PanelStateManager::setCEState()
...
+}
+
+types::Binary PanelStateManager::getEnabledFunctionsList()
+{
+    types::Binary inputFunctions = {21, 22, 34, 65, 66, 67, 68, 69, 70};
+
+    types::Binary enabledFunctions;
+    enabledFunctions.reserve(inputFunctions.size());
+
+    for (auto& itr : inputFunctions)
```
Comment (SunnySrivastava1984) [inline src/panel_state_manager.cpp:1135]: Change the name "itr" as it is not iterator here. It is the value in the vector.

**Review by SunnySrivastava1984 — CHANGES_REQUESTED** [review]

**Review by SunnySrivastava1984 — CHANGES_REQUESTED** [review]

File: src/panel_state_manager.cpp (hunk 41 above)
Comment (PriyangaRamasamy) [reply src/panel_state_manager.cpp:1121]: done

File: include/bus_handler.hpp (hunk 40 above)
Comment (PriyangaRamasamy) [reply include/bus_handler.hpp:74]: will follow properly from next time

File: src/panel_state_manager.cpp (hunk 42 above)
Comment (PriyangaRamasamy) [reply src/panel_state_manager.cpp]: done

**Review by SunnySrivastava1984 — APPROVED** [review]

**Review by jinuthomas — APPROVED** [review]
